### PR TITLE
Fix the API of bce loss and add functional API binary_cross_entropy

### PR DESCRIPTION
**Pull request by ZHUI**
### PR types

Others
### PR changes

APIs
### Describe

fix the API of bce loss and add functional API binary_cross_entropy.

### CODE REVIEW

File: python/paddle/nn/layer/loss.py [hunk 1]
```diff
@@ -340,90 +341,97 @@ def forward(self, input, label):
 
 class BCELoss(fluid.dygraph.Layer):
     """
-	:alias_main: paddle.nn.BCELoss
-	:alias: paddle.nn.BCELoss,paddle.nn.layer.BCELoss,paddle.nn.layer.loss.BCELoss
-
     This interface is used to construct a callable object of the ``BCELoss`` class.
-    The BCELoss layer measures the binary_cross_entropy loss between input predictions 
-    and target labels. The binary_cross_entropy loss can be described as:
+    The BCELoss layer measures the binary_cross_entropy loss between input predictions ``x``
```
Comment (wawltor) [inline python/paddle/nn/layer/loss.py]: x->input

File: python/paddle/nn/layer/loss.py [hunk 2]
```diff
@@ -340,90 +341,97 @@ def forward(self, input, label):
...
+    and target labels ``label`` . The binary_cross_entropy loss can be described as:
 
     If :attr:`weight` is set, the loss is:
 
     .. math::
-        Out = -1 * weight * (label * log(input) + (1 - label) * log(1 - input))
+        Out = -1 * weight * (label * log(x) + (1 - label) * log(1 - x))
```
Comment (wawltor) [inline python/paddle/nn/layer/loss.py]: x->input, 同下

File: python/paddle/nn/layer/loss.py [hunk 3]
```diff
@@ -340,90 +341,97 @@ def forward(self, input, label):
...
     should be numbers between 0 and 1.
 
-    The shape of input predictions and target labels are [N, *], where N is batch_size and `*` 
-    means any number of additional dimensions. If ``reduction`` is ``'none'``, the shape of 
-    output is scalar, else the shape of output is same as input.
-
     Parameters:
-        weight (Variable, optional): A manual rescaling weight given to the loss of each 
+        weight (Tensor, optional): A manual rescaling weight given to the loss of each
             batch element. If given, has to be a Variable of size nbatch and the data type
```
Comment (wawltor) [inline python/paddle/nn/layer/loss.py]: Variable -> Tensor

File: python/paddle/nn/layer/loss.py [hunk 4]
```diff
@@ -340,90 +341,97 @@ def forward(self, input, label):
...
             If :attr:`reduction` is ``'none'``, the unreduced loss is returned;
-            If :attr:`reduction` is ``'mean'``, the reduced mean loss is returned; 
+            If :attr:`reduction` is ``'mean'``, the reduced mean loss is returned;
             If :attr:`reduction` is ``'sum'``, the summed loss is returned.
             Default is ``'mean'``.
+        name (str, optional): Name for the operation (optional, default is None).
+            For more information, please refer to :ref:`api_guide_Name`.
+
+    Shape:
+        x (Tensor): 2-D tensor with shape: (N, *), N is batch_size, `*` means
```
Comment (wawltor) [inline python/paddle/nn/layer/loss.py]: x->input

File: python/paddle/nn/layer/loss.py [hunk 5]
```diff
@@ -340,90 +341,97 @@ def forward(self, input, label):
...
+            place = paddle.static.CPUPlace()
+            exe = paddle.static.Executor(place)
             exe.run(fluid.default_startup_program())
-    
-            input_data = np.array([0.5, 0.6, 0.7]).astype("float32")
-            label_data = np.array([1.0, 0.0, 1.0]).astype("float32")
+
             output_data = exe.run(fluid.default_main_program(),
-                    feed={"input":input_data, "label":label_data},
+                    feed={"x":x_data, "label":label_data},
```
Comment (wawltor) [inline python/paddle/nn/layer/loss.py]: x->input

File: python/paddle/fluid/tests/unittests/test_bce_loss.py [hunk 6]
```diff
@@ -19,90 +19,105 @@
...
-                weight=fluid.dygraph.to_variable(weight_np))
-            dy_res = bce_loss(
-                fluid.dygraph.to_variable(input_np),
-                fluid.dygraph.to_variable(label_np))
-            dy_result = dy_res.numpy()
-
-        expected = np.mean(-1 * weight_np *
-                           (label_np * np.log(input_np) +
-                            (1. - label_np) * np.log(1. - input_np)))
+        for red in ['sum', 'none']:
```
Comment (wawltor) [inline python/paddle/fluid/tests/unittests/test_bce_loss.py]: 加一个mean

File: python/paddle/nn/layer/loss.py [hunk 7]
```diff
@@ -333,90 +334,97 @@ def forward(self, x, label):
...
-            # declarative mode
             import paddle.fluid as fluid
             import numpy as np
             import paddle
-            input = fluid.data(name="input", shape=[3, 1], dtype='float32')
-            label = fluid.data(name="label", shape=[3, 1], dtype='float32')
+            input_data = np.array([0.5, 0.6, 0.7]).astype("float32")
+            label_data = np.array([1.0, 0.0, 1.0]).astype("float32")
+
+            # imperative mode
```
Comment (joey12300) [inline python/paddle/nn/layer/loss.py]: Don't reflect the word `imperative`.

File: python/paddle/nn/layer/loss.py [hunk 8]
```diff
@@ -333,90 +334,97 @@ def forward(self, x, label):
...
+            input = paddle.to_variable(input_data)
+            label = paddle.to_variable(label_data)
             bce_loss = paddle.nn.loss.BCELoss()
             output = bce_loss(input, label)
-            place = fluid.CPUPlace()
-            exe = fluid.Executor(place)
+            print(output.numpy())  # [0.65537095]
+            paddle.enable_static()
+
+            # static mode
```
Comment (joey12300) [inline python/paddle/nn/layer/loss.py]: No need to show static graph demo code.

File: python/paddle/fluid/tests/unittests/test_bce_loss.py [hunk 9]
```diff
@@ -19,48 +19,77 @@
 from op_test import OpTest
 
 
+def test_static(place, input_np, label_np, red='mean', weight_np=None):
```
Comment (joey12300) [inline python/paddle/fluid/tests/unittests/test_bce_loss.py]: red->reduction

File: python/paddle/fluid/tests/unittests/test_bce_loss.py [hunk 10]
```diff
@@ -19,48 +19,77 @@
...
+        expected = np.mean(expected)
+    elif red == 'sum':
+        expected = np.sum(expected)
+    else:
+        expected = expected
+
+    return expected
+
+
 class TestBCELoss(unittest.TestCase):
```
Comment (joey12300) [inline python/paddle/fluid/tests/unittests/test_bce_loss.py:140]: miss reduction Value Error test case

File: python/paddle/nn/layer/loss.py (hunk 4 above)
Comment (ZHUI) [reply python/paddle/nn/layer/loss.py]: done

File: python/paddle/nn/layer/loss.py (hunk 3 above)
Comment (ZHUI) [reply python/paddle/nn/layer/loss.py]: done

File: python/paddle/nn/layer/loss.py (hunk 7 above)
Comment (ZHUI) [reply python/paddle/nn/layer/loss.py]: done

File: python/paddle/nn/layer/loss.py (hunk 8 above)
Comment (ZHUI) [reply python/paddle/nn/layer/loss.py]: there exist static mode demo before. I think keep it will be batter

File: python/paddle/fluid/tests/unittests/test_bce_loss.py (hunk 10 above)
Comment (ZHUI) [reply python/paddle/fluid/tests/unittests/test_bce_loss.py:140]: done

File: python/paddle/fluid/tests/unittests/test_bce_loss.py (hunk 9 above)
Comment (ZHUI) [reply python/paddle/fluid/tests/unittests/test_bce_loss.py]: done

**Comment by LutaoChu** [comment]
> ### PR types
> Others
> 
> ### PR changes
> Docs
> 
> ### Describe
> PREVIEW THE DOCS, DON'T MERGE!

describe记得修改

File: python/paddle/fluid/tests/unittests/test_bce_loss.py [hunk 11]
```diff
@@ -19,48 +19,79 @@
...
-                    expected = np.sum(expected)
-                else:
-                    expected = expected
+            for reduction in reductions:
+                static_result = test_static(place, input_np, label_np,
+                                            reduction)
+                dy_result = test_dygraph(place, input_np, label_np, reduction)
+                expected = calc_bceloss(input_np, label_np, reduction)
                 self.assertTrue(np.allclose(static_result, expected))
                 self.assertTrue(np.allclose(static_result, dy_result))
```
Comment (LutaoChu) [inline python/paddle/fluid/tests/unittests/test_bce_loss.py:156]: 如果上面和下面都pass，这个判断没有必要了

File: python/paddle/fluid/tests/unittests/test_bce_loss.py [hunk 12]
```diff
@@ -69,44 +100,38 @@ def test_BCELoss_weight(self):
...
-                            (1. - label_np) * np.log(1. - input_np)))
+        for reduction in ['sum', 'mean', 'none']:
+            static_result = test_static(
+                place, input_np, label_np, reduction, weight_np=weight_np)
+            dy_result = test_dygraph(
+                place, input_np, label_np, reduction, weight_np=weight_np)
+            expected = calc_bceloss(
+                input_np, label_np, reduction, weight_np=weight_np)
+            self.assertTrue(np.allclose(static_result, expected))
+            self.assertTrue(np.allclose(static_result, dy_result))
```
Comment (LutaoChu) [inline python/paddle/fluid/tests/unittests/test_bce_loss.py:182]: 同上，可以删除

File: python/paddle/fluid/tests/unittests/test_bce_loss.py [hunk 13]
```diff
@@ -69,44 +100,38 @@ def test_BCELoss_weight(self):
...
+        input_np = np.random.random(size=(2, 3, 4, 10)).astype(np.float64)
+        label_np = np.random.random(size=(3, 4, 10)).astype(np.float64)
+        place = fluid.CUDAPlace(0) if fluid.core.is_compiled_with_cuda(
+        ) else fluid.CPUPlace()
+
+        static_result = test_static(place, input_np, label_np)
+        dy_result = test_dygraph(place, input_np, label_np)
+        expected = calc_bceloss(input_np, label_np)
         self.assertTrue(np.allclose(static_result, expected))
         self.assertTrue(np.allclose(static_result, dy_result))
```
Comment (LutaoChu) [inline python/paddle/fluid/tests/unittests/test_bce_loss.py:203]: 同上

File: python/paddle/nn/layer/loss.py [hunk 14]
```diff
@@ -333,90 +334,96 @@ def forward(self, x, label):
...
     .. math::
         Out = -1 * weight * (label * log(input) + (1 - label) * log(1 - input))
+
     If :attr:`weight` is None, the loss is:
 
     .. math::
         Out = -1 * (label * log(input) + (1 - label) * log(1 - input))
 
-    If :attr:`reduction` set to ``'none'``, the unreduced loss is:
+    If :attr:`reduction` set to ``'none'``, the interface will return the original loss `Out`.
```
Comment (LutaoChu) [inline python/paddle/nn/layer/loss.py:353]: interface用法会不会有点奇怪，改为operation？

File: python/paddle/nn/functional/loss.py [hunk 15]
```diff
@@ -72,6 +74,143 @@
...
+            N is batch_size, `*` means number of additional dimensions. The ``input``
+            should always be the output of sigmod.  Available dtype is float32, float64.
+        label (Tensor): The target labels tensor. 2-D tensor with the same shape as
+            ``input``. The target labels which values should be numbers between 0 and 1.
+            Available dtype is float32, float64.
+        weight (Tensor, optional): A manual rescaling weight given to the loss of each
+            batch element. If given, has to be a Tensor of size nbatch and the data type
+            is float32, float64. Default is ``'None'``.
+        reduction (str, optional): Indicate how to average the loss by batch_size,
+            the candicates are ``'none'`` | ``'mean'`` | ``'sum'``.
```
Comment (wawltor) [inline python/paddle/nn/functional/loss.py:120]: | 要确定改成什么符号合适？

File: python/paddle/nn/functional/loss.py [hunk 16]
```diff
@@ -72,6 +74,143 @@
...
+
+
+    Returns:
+        output (Tensor): If ``reduction`` is ``'none'``, the shape of output is
+            same as ``input`` , else the shape of output is scalar.
+
+    Examples:
+        .. code-block:: python
+
+            import paddle.fluid as fluid
```
Comment (wawltor) [inline python/paddle/nn/functional/loss.py]: 去掉这行

File: python/paddle/fluid/tests/unittests/test_bce_loss.py (hunk 11 above)
Comment (ZHUI) [reply python/paddle/fluid/tests/unittests/test_bce_loss.py:156]: 建议保留。例如误差范围为 d， static_result  = expected + d, dy_result = expected - d。此时，static_result - dy_result = 2d，可能超出误差范围。

Comment (LutaoChu) [reply python/paddle/fluid/tests/unittests/test_bce_loss.py:156]: 也可以

**Review by wawltor — APPROVED** [review]
LGTM

File: python/paddle/nn/functional/loss.py [hunk 17]
```diff
@@ -73,6 +75,142 @@
...
+            input_data = np.array([0.5, 0.6, 0.7]).astype("float32")
+            label_data = np.array([1.0, 0.0, 1.0]).astype("float32")
+
+            paddle.disable_static()
+            input = paddle.to_tensor(input_data)
+            label = paddle.to_tensor(label_data)
+            output = paddle.nn.functional.binary_cross_entropy(input, label)
+            print(output.numpy())  # [0.65537095]
+            paddle.enable_static()
+
```
Comment (TCChenlong) [inline python/paddle/nn/functional/loss.py:147]: 这一行多余吧？
